File: cpp/tree.cpp
#include <iostream>
#include <set>
#include <map>
#include <deque>
#include <queue>
#include <vector>
#include <string>
#include <sstream>
#include <numeric>
#include <algorithm>
#include <functional>
using namespace std;

#define null -1
#define VI(...) std::vector<int>({__VA_ARGS__})
#define VS(...) std::vector<string>({__VA_ARGS__})

template <typename T>
void printVec(const vector<T> &v)
{
    cout << "[";
    for (auto x : v)
        cout << x << ",";
    cout << "]" << endl;
}

template <typename T, short s>
vector<T> &&genVec(const T v[s])
{
    vector<T> vec(v);
    return move(vec);
}

// -----

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

// v: {1, 2, null, 3} => {1, 2, -1, 3}
TreeNode *genTree(const vector<int> &v)
{
    if (v.empty())
        return nullptr;
    deque<TreeNode *> q;
    auto root = new TreeNode(v[0]);
    q.push_back(root);
    for (int i = 1; i < v.size(); ++i)
    {
        TreeNode *l = nullptr, *r = nullptr;
        if (v[i] >= 0)
            l = new TreeNode(v[i]);
        if (i + 1 < v.size())
        {
            i++;
            if (v[i] >= 0)
                r = new TreeNode(v[i]);
        }
        q.front()->left = l;
        q.front()->right = r;
        q.pop_front();
        if (l)
            q.push_back(l);
        if (r)
            q.push_back(r);
    }
    return root;
};

void printTree(const TreeNode *root)
{
    vector<const TreeNode *> vt;
    vt.push_back(root);
    while (!vt.empty())
    {
        vector<const TreeNode *> new_vt;
        int non_null_count = 0;
        for (auto t : vt)
            if (t)
                non_null_count++;
        if (non_null_count == 0)
            break;
        for (auto t : vt)
        {
            if (t && t->val >= 0)
            {
                cout << t->val << ",";
                new_vt.push_back(t->left);
                new_vt.push_back(t->right);
                non_null_count++;
            }
            else
                cout << "null,";
        }
        vt.swap(new_vt); // vt = new_vt
    }
    cout << endl;
}

void printTreePretty(const TreeNode *root)
{
    deque<const TreeNode *> q;
    q.push_back(root);
    while (!q.empty() /* never reaches */)
    {
        int lvlcnt = q.size();
        bool nextlvlnoempty = false;
        cout << "| ";

        for (int i = 0; i < lvlcnt; ++i)
        {
            auto *node = q.front();
            q.pop_front();
            if (!node)
            {
                cout << "- ";
                q.push_back(nullptr);
                q.push_back(nullptr);
            }
            else
            {
                cout << node->val << " ";
                q.push_back(node->left);
                q.push_back(node->right);
                nextlvlnoempty |= node->left || node->right;
            }
        }
        cout << endl;
        if (!nextlvlnoempty)
            break;
    }
}

File: rust/tree.rs
use std::{cell::RefCell, collections::VecDeque, rc::Rc};

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

fn gen_tree(arr: &[i32]) -> Option<Rc<RefCell<TreeNode>>> {
    if arr.len() == 0 {
        return None;
    }
    let mut q = VecDeque::from([Some(Rc::new(RefCell::new(TreeNode::new(arr[0]))))]);

    // Clone is acceptable as we're just increasing RefCount
    let root = match &q[0] {
        Some(x) => Some(Rc::clone(x)),
        _ => unreachable!(),
    };

    // Now we dig into queue and input array to build the tree
    let mut aiter = arr.iter().peekable();
    aiter.next(); // consume first one as it's already used
    while !q.is_empty() && aiter.peek().is_some() {
        let width_len = q.len();
        (0..width_len).for_each(|_| {
            let node = q.pop_front().unwrap();
            match &node {
                Some(x) => {
                    let left = if let Some(n) = aiter.next() {
                        if *n >= 0 {
                            Some(Rc::new(RefCell::new(TreeNode::new(*n))))
                        } else {
                            None
                        }
                    } else {
                        None
                    };
                    let right = if let Some(n) = aiter.next() {
                        if *n >= 0 {
                            Some(Rc::new(RefCell::new(TreeNode::new(*n))))
                        } else {
                            None
                        }
                    } else {
                        None
                    };
                    // clone is okay as it's RC
                    q.push_back(left.clone());
                    q.push_back(right.clone());
                    let mut xmut = x.borrow_mut();
                    xmut.left = left;
                    xmut.right = right;
                }
                _ => {
                    // just skip two arr inputs
                    aiter.next();
                    aiter.next();
                    q.push_back(None);
                    q.push_back(None);
                }
            }
        });
    }

    root
}

fn print_tree(t: Option<Rc<RefCell<TreeNode>>>) {
    let mut v = VecDeque::from([t]);
    while !v.is_empty() {
        let curr_len = v.len();
        let mut alive_elem_cnt = 0;
        let curr_level = (0..curr_len)
            .map(|_| {
                match v.pop_front() {
                    Some(Some(n)) => {
                        alive_elem_cnt += 1;
                        let n = n.borrow();
                        // Add all child in the queue
                        // Clone is cost-safe as RC only increases RefCount
                        v.push_back(match &n.left {
                            Some(x) => Some(Rc::clone(&x)),
                            _ => None,
                        });
                        v.push_back(match &n.right {
                            Some(x) => Some(Rc::clone(&x)),
                            _ => None,
                        });
                        n.val.to_string()
                    }
                    _ => {
                        v.push_back(None);
                        v.push_back(None);
                        ".".to_string()
                    }
                }
            })
            .collect::<Vec<_>>();
        if alive_elem_cnt == 0 {
            break;
        }
        println!("ㅏ {:?}", curr_level);
    }
}

fn main() {
    // empty tree
    let sample_tree = gen_tree(&[]);
    print_tree(sample_tree);
    println!("---");

    // full binary tree
    let sample_tree = gen_tree(&[1, 2, 3, 4, 5, 6, 7]);
    print_tree(sample_tree);
    println!("---");

    // plain, sparse binary tree
    let sample_tree = gen_tree(&[1, -1, 2, -1, -1, 3]);
    print_tree(sample_tree);
    println!("---");

    println!("Hello, world!");
}
